src/mono/mono/metadata/assembly.c

```diff
@@ -4317,14 +4318,14 @@ assembly_binding_info_parsed (MonoAssemblyBindingInfo *info, void *user_data)
 			return;
 	}
 
-	info_copy = (MonoAssemblyBindingInfo *)mono_mempool_alloc0 (domain->mp, sizeof (MonoAssemblyBindingInfo));
+	info_copy = (MonoAssemblyBindingInfo *)mono_mempool_alloc0 (memory_manager->mp, sizeof (MonoAssemblyBindingInfo));
```

Comment:
This needs locking on the memory manager. In general, it would be better to add functions to allocate from the memory manager mp which do the locking themselves.

Reply:
This worked because the parent locked the domain lock. Probably to save locking on every allocation, since there are a lot of them. Would be better to add _unlocked versions of the alloc function to avoid direct access to ->mp and make this more explicit.

Reply:
> it would be better to add functions to allocate from the memory manager mp which do the locking themselves

I have those and just didn't include them in this PR since it made the diff a bit messier. If you think they're fine here I can push and update some of these.